src/backend/utils/adt/gp_optimizer_functions.c

```diff
@@ -16,47 +16,77 @@
 #include "funcapi.h"
 #include "utils/builtins.h"
 
-extern Datum EnableXform(PG_FUNCTION_ARGS);
+typedef Datum (*gp_opt_version_func) (void);
+
+/*
+ * A stub function to call if no optimizer function is found.
+ */
+static Datum
+gp_optimizer_function_stub(PG_FUNCTION_ARGS)
+{
+	return CStringGetTextDatum("gp_orca is not added to 'shared_preload_libraries'");
+}
+
+/*
+ * Loads optimizer function from a shared library. If library is not presented
+ * or doesn't contain the requested function or any of its dependencies,
+ * returns the stub function.
+ */
+static PGFunction
+gp_optimizer_load_function(char *funcname)
+{
+	volatile PGFunction func = NULL;
+	int32		savedInterruptHoldoffCount = InterruptHoldoffCount;
+	MemoryContext oldcontext = CurrentMemoryContext;
+
+	PG_TRY();
+	{
+		func = load_external_function("$libdir/gp_orca", funcname, false, NULL);
+	}
+	PG_CATCH();
+	{
+		MemoryContextSwitchTo(oldcontext);
+		InterruptHoldoffCount = savedInterruptHoldoffCount;
+		FlushErrorState();
+	}
+	PG_END_TRY();
+
+	if (NULL == func)
+		func = gp_optimizer_function_stub;
```

Comment:
how about to change behavior to more logical
```suggestion
		elog(ERROR, "gp_orca is not added to 'shared_preload_libraries'");
```

Reply:
Current tests expect that the functions will not end with an error if Orca is not enabled.

Btw, tests were updated as the message had changed.